walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,79 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
+
+  // has one slot for every minute of a day
+  // The minute is true if all attendees are available in that minute, false
+  // otherwise
+  private boolean[] availability = new boolean[TimeRange.END_OF_DAY + 1];
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Arrays.fill(availability,true);
+
+    for (Event event : events) {
+      if (isEventRelevant(event, request.getAttendees())) {
+        recordBusyTime(event.getWhen());
+      }
+    }
+
+    Collection<TimeRange> availableTimes = new LinkedList<TimeRange>();
+
+    int startIndex = 0;
+
+    while (startIndex <= TimeRange.END_OF_DAY) {
+        if (!availability[startIndex]) {
+            startIndex ++;
+            continue;
+        }
+        
+        int endIndex = startIndex;
+
+        while (endIndex <= TimeRange.END_OF_DAY) {
+            if (!availability[endIndex]) {
+                break;
+            }
+            endIndex ++;
+        }
+
+
+        if (endIndex == TimeRange.END_OF_DAY && availability[endIndex]) {
+            endIndex ++;
+        }
+
+        if (endIndex - startIndex >= request.getDuration()) {
+            TimeRange availableTime = TimeRange.fromStartEnd(startIndex, endIndex, false);
+            availableTimes.add(availableTime);
+        }
+
+        startIndex  = endIndex;
```

Comment:
If the above TimeRange represents an inclusive window (ie; from the start of startIndex to the end of endIndex), then the next iteration of the loop should start at endIndex + 1, otherwise it'll always perform an extra check against the index we just ended on?

Reply:
I think that the time range has an exclusive endpoint, because of the `false` parameter used in it's constructor in line 63